<a href="https://colab.research.google.com/github/fralfaro/ICS40125/blob/main/docs/projects/project_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Analítica de Recursos Humanos: Determinantes Salariales y Modelación de Compensaciones

## Contexto y Objetivo

Una firma multinacional de consultoría ha sido contratada por **TalentCo**, empresa con operaciones en múltiples regiones y departamentos, para analizar su estructura de compensaciones. El equipo de Recursos Humanos sospecha que los salarios no están siendo asignados de manera consistente: algunos empleados se sienten subpagados en relación con pares de experiencia similar, mientras que otros en el mismo departamento perciben montos significativamente distintos según su región o nivel educacional.

Su rol es actuar como **Analista de Datos** en la firma consultora. Utilizará Python, pruebas de hipótesis estadísticas y regresión lineal para identificar los principales determinantes del salario de los empleados, y entregará recomendaciones basadas en datos a la dirección de TalentCo.

Al finalizar este proyecto, usted será capaz de:

- Realizar análisis exploratorio de datos (EDA) sobre datos de Recursos Humanos.
- Aplicar pruebas de hipótesis para validar supuestos de negocio.
- Manejar variables categóricas mediante técnicas de codificación.
- Construir y evaluar modelos de regresión con OLS y scikit-learn.
- Verificar los supuestos de regresión (multicolinealidad, normalidad, homocedasticidad).
- Comunicar resultados en un informe orientado al negocio.

## Conjunto de Datos: `employee_compensation.csv`

Disponible en la sección de materiales del curso. Contiene registros de **1.500 empleados** distribuidos en distintos departamentos, regiones y niveles de cargo.

| Variable | Tipo | Descripción |
|---|---|---|
| `Salary` | Numérica (USD/año) | Salario bruto anual — **variable objetivo** |
| `Years_Experience` | Numérica | Total de años de experiencia profesional |
| `Age` | Numérica | Edad del empleado |
| `Performance_Score` | Numérica (1.0–5.0) | Calificación de desempeño anual más reciente |
| `Education_Level` | Categórica | Bachelor, Master, PhD |
| `Department` | Categórica | Engineering, Sales, Marketing, HR, Finance |
| `Region` | Categórica | North, South, East, West |
| `Job_Level` | Categórica | Junior, Mid, Senior |
| `Gender` | Categórica | Male, Female |

## Parte 1: Análisis Exploratorio de Datos (EDA)

1. Cargue el dataset en un DataFrame de pandas.
2. Muestre las primeras 10 filas e inspeccione los tipos de datos.
3. Verifique valores faltantes y duplicados; reporte los hallazgos.
4. Calcule estadísticas descriptivas (media, mediana, desviación estándar, mínimo, máximo) para todas las variables numéricas.
5. Elabore al menos **tres** de las siguientes visualizaciones:
   - Histograma del salario.
   - Diagrama de caja (*box plot*) del salario por nivel de cargo.
   - Diagrama de dispersión de años de experiencia vs. salario.
   - Gráfico de barras del salario promedio por departamento.
   - Mapa de calor de correlaciones entre variables numéricas.



# **1. Cargue un dataset en un DataFrame de pandas**

In [26]:
import pandas as pd

# cargar datos
salary = 'https://raw.githubusercontent.com/fralfaro/ICS40125/main/docs/projects/data/employee_compensation.csv'
df_salary = pd.read_csv(salary, sep="," )


# **2. Muestre las primeras filas e inspeccione los tipos de datos**

In [27]:
df_salary.head(10)

,Salary,Years_Experience,Age,Performance_Score,Education_Level,Department,Region,Job_Level,Gender
0,69181.0,3.0,22.0,3.5,Bachelor,Sales,East,Junior,Male
1,88466.0,5.0,23.0,2.8,bachelor,FINANCE,West,Junior,Female
2,76112.0,11.0,29.0,1.6,Master,HR,West,Mid,Female
3,NaN,15.0,37.0,NaN,Bachelor,Marketing,South,Senior,Male
4,NaN,6.0,28.0,4.5,Bachelor,Engineering,North,Mid,Female
5,81246.0,4.0,21.0,2.8,Bachelor,Engineering,West,Junior,Male
6,71494.0,4.0,23.0,2.4,Bachelor,Engineering,South,Junior,Male
7,78189.0,3.0,21.0,3.5,Master,Finance,East,Junior,NaN
8,67204.0,0.0,21.0,3.5,Master,Marketing,East,Junior,Female
9,59776.0,4.0,24.0,NaN,Bachelor,HR,East,Junior,Female


In [28]:
df_salary.dtypes

,0
Salary,float64
Years_Experience,float64
Age,float64
Performance_Score,float64
Education_Level,object
Department,object
Region,object
Job_Level,object
Gender,object


El tipo de dato para el salario bruto anual (salary), total de años de experiencia profesional (Years_Experience), edad del empleado (Age) y calificación de desempeño anual más reciente (performance_Score) son de tipo numérica que permite decimales. El resto de variables respecto al nivel eduacional (Education_Level), departamento (Department	Categórica), región (Region), nivel en la empresa (Job_Level) y género (Gender)	corresponden a variables de cadena de texto (strings).

# **3. Verifique valores faltantes y duplicados; reporte los hallazgos.**

In [29]:
df_salary.index

RangeIndex(start=0, stop=10100, step=1)

In [30]:
df_salary.isnull().sum()

,0
Salary,203
Years_Experience,405
Age,303
Performance_Score,603
Education_Level,506
Department,305
Region,401
Job_Level,0
Gender,501


In [31]:
df_salary_null= df_salary.isnull().sum()

porcentaje_null = (100*df_salary_null)/10100
porcentaje_null

,0
Salary,2.009901
Years_Experience,4.009901
Age,3.000000
Performance_Score,5.970297
Education_Level,5.009901
Department,3.019802
Region,3.970297
Job_Level,0.000000
Gender,4.960396


**3.1 Valores Faltantes (nulos)**
Respecto a esto podemos dar cuenta que las variables con más missing values son Performance_Score, Educational_Level y Gender.
La base de datos posee 10.100 valores por lo que el porcentaje de missing values correpsondiente a cada columna se encuentra a continuación:

| Columna            | Porcentaje de Valores Faltantes (%) |
|:-------------------|:------------------------------------|
| Salary             | 2.01                                |
| Years_Experience   | 4.01                                |
| Age                | 3.00                                |
| Performance_Score  | 5.97                                |
| Education_Level    | 5.01                                |
| Department         | 3.02                                |
| Region             | 3.97                                |
| Job_Level          | 0.00                                |
| Gender             | 4.96     



In [45]:
df_salary_imputed = df_salary.copy()

for col in ['Salary', 'Years_Experience', 'Age', 'Performance_Score']:
    if df_salary_imputed[col].isnull().any():
        median_value = df_salary_imputed[col].median()
        df_salary_imputed[col] = df_salary_imputed[col].fillna(median_value)

for col in ['Education_Level', 'Department', 'Region', 'Gender']:
    if df_salary_imputed[col].isnull().any():
        mode_value = df_salary_imputed[col].mode()[0]
        df_salary_imputed[col] = df_salary_imputed[col].fillna(mode_value)

print("\nValores faltantes después del tratamiento en df_salary_imputed:")
print(df_salary_imputed.isnull().sum())


Valores faltantes después del tratamiento en df_salary_imputed:
Salary               0
Years_Experience     0
Age                  0
Performance_Score    0
Education_Level      0
Department           0
Region               0
Job_Level            0
Gender               0
dtype: int64


Para asegurar la completitud de los datos y mantener la robustez del análisis, se procedió a la imputación de los valores faltantes en el DataFrame df_salary_imputed. Los valores nulos en las columnas numéricas 'Salary', 'Years_Experience', 'Age' y 'Performance_Score' fueron reemplazados con sus respectivas medianas (91701.0, 6.0, 29.0 y 3.0, respectivamente), una estrategia elegida por su menor sensibilidad a valores atípicos. Para las variables categóricas 'Education_Level', 'Department', 'Region' y 'Gender', los valores faltantes se imputaron con la moda de cada columna ('Bachelor', 'Engineering', 'South' y 'Female', respectivamente). Tras esta operación, se confirmó que el DataFrame df_salary_imputed no contiene valores nulos en ninguna de sus columnas.

**3.2 Valores Duplicado**


In [49]:
df_salary_imputed.duplicated().sum()

np.int64(97)

In [51]:
df_droped = df_salary_imputed.drop_duplicates()
df_droped.duplicated().sum()

np.int64(0)

Posterior a la imputación de valores faltantes, se identificó la presencia de 97 registros duplicados en el conjunto de datos `df_salary_imputed`. Esta situación es una consecuencia esperada del proceso de imputación, donde filas previamente diferenciadas por la ausencia de datos pueden converger a valores idénticos al ser rellenadas con estadísticas comunes (media o moda).

Para garantizar la integridad y unicidad de cada observación en el análisis subsiguiente, estos registros duplicados han sido eliminados, resultando en un DataFrame (`df_droped`) que ahora no contiene duplicados. Esta acción es fundamental para evitar sesgos en el modelado y asegurar que cada empleado esté representado una única vez en el análisis.

In [53]:
#TERMINARRRR
# Filtrar 'Years_Experience': solo valores mayores o iguales a 0
df_Salary_final = df_droped[df_droped['Years_Experience'] >= 0]

# Filtrar 'Age': solo valores entre 18 y 100 (inclusive)
df_Salary_final = df_Salary_final[
    (df_Salary_final['Age'] >= 18) & (df_Salary_final['Age'] <= 100)
]

# Filtrar 'Performance_Score': solo valores entre 1 y 5 (inclusive)
df_Salary_final = df_Salary_final[
    (df_Salary_final['Performance_Score'] >= 1) & (df_Salary_final['Performance_Score'] <= 5)
]

print("Dimensiones del DataFrame original (df_droped):")
print(df_droped.shape)
print("\nDimensiones del DataFrame final (df_Salary_final) después de limpiar valores imposibles:")
print(df_Salary_final.shape)

print("\nPrimeras 5 filas del df_Salary_final:")
display(df_Salary_final.head())

Dimensiones del DataFrame original (df_droped):
(10003, 9)

Dimensiones del DataFrame final (df_Salary_final) después de limpiar valores imposibles:
(9935, 9)

Primeras 5 filas del df_Salary_final:


,Salary,Years_Experience,Age,Performance_Score,Education_Level,Department,Region,Job_Level,Gender
0,69181.0,3.0,22.0,3.5,Bachelor,Sales,East,Junior,Male
1,88466.0,5.0,23.0,2.8,bachelor,FINANCE,West,Junior,Female
2,76112.0,11.0,29.0,1.6,Master,HR,West,Mid,Female
3,91701.0,15.0,37.0,3.0,Bachelor,Marketing,South,Senior,Male
4,91701.0,6.0,28.0,4.5,Bachelor,Engineering,North,Mid,Female


# **4. Calcule estadísticas descriptivas para todas las variables numéricas.**


In [52]:
df_droped.describe()

,Salary,Years_Experience,Age,Performance_Score
count,10003.000000,10003.000000,10003.000000,10003.000000
mean,95687.480256,6.791463,29.455363,3.026362
std,30351.502663,5.579057,6.750686,1.759577
min,5342.000000,-10.000000,1.000000,-1.000000
25%,77300.000000,3.000000,25.000000,2.100000
50%,91701.000000,6.000000,29.000000,3.000000
75%,111451.000000,10.000000,33.000000,3.900000
max,394249.000000,78.000000,119.000000,99.000000




## Parte 2: Pruebas de Hipótesis

**Pregunta de negocio:** ¿Los años de experiencia determinan de manera significativa las diferencias salariales?

1. Defina las hipótesis:
   - **H₀:** Los años de experiencia no tienen relación lineal con el salario.
   - **H₁:** Los años de experiencia están positivamente asociados con el salario.

2. Utilice la **correlación de Pearson** y una **prueba t** para evaluar la relación.

3. Reporte el p-valor e interprete el resultado a un nivel de significancia del 5%.

4. **(Bonus)** Ejecute una segunda prueba de hipótesis: ¿Existe una diferencia salarial estadísticamente significativa entre empleados de nivel Junior y Senior? Utilice una prueba t de dos muestras independientes.



In [35]:
#fixme



## Parte 3: Regresión Lineal con Variables Categóricas

### Paso 1: Justificación de variables (antes de codificar)

Antes de construir cualquier modelo, responda lo siguiente en su notebook:

- ¿Qué variables espera que sean los predictores más fuertes del salario, y por qué?
- ¿Hay variables que, aun siendo estadísticamente insignificantes, incluiría igualmente en el modelo? Justifique desde una perspectiva de negocio o teórica.
- ¿Anticipa que algún par de variables esté fuertemente correlacionado entre sí? ¿Cuáles, y por qué? ¿Qué problemas podría causar eso?

> Este paso debe responderse en texto libre, sin código. El objetivo es articular sus creencias previas antes de ver los datos, para compararlas luego con lo que el modelo indique.

### Paso 2: Preparación de los datos

- Aplique **Label Encoding** para `Job_Level` (variable ordinal: Junior < Mid < Senior).
- Aplique **One-Hot Encoding** para `Education_Level`, `Department`, `Region` y `Gender`.
- Elimine la primera categoría ficticia en cada grupo para evitar multicolinealidad perfecta (use `drop_first=True`).

### Paso 3: Construcción del modelo

1. Divida el dataset en conjuntos de entrenamiento (80%) y prueba (20%). Use `random_state=42`.
2. Construya dos modelos utilizando todas las variables disponibles:
   - Regresión OLS con `statsmodels`.
   - Regresión lineal con `scikit-learn`.

### Paso 4: Evaluación del modelo

Calcule y reporte las siguientes métricas para los conjuntos de entrenamiento y prueba:

| Métrica | Descripción |
|---|---|
| MAE | Error Absoluto Medio |
| MAPE | Error Porcentual Absoluto Medio |
| RMSE | Raíz del Error Cuadrático Medio |
| R² | Coeficiente de determinación |
| R² ajustado | R² penalizado por el número de predictores |

Compare el resumen OLS (coeficientes, p-valores, R²) con los resultados de scikit-learn.



In [36]:
#fixme



## Parte 4: Análisis y Refinamiento del Modelo

> **Nota importante:** El objetivo de esta sección no es aplicar reglas mecánicas. No existe un umbral universal que indique cuándo incluir o excluir una variable. Toda decisión debe argumentarse desde el contexto del problema.

### Paso 1: Interpretación de los p-valores de los coeficientes

Examine los p-valores del resumen OLS y responda lo siguiente para cada variable con p-valor > 0,05:

- ¿La variable es estadísticamente insignificante porque no tiene efecto real, o podría haber otra explicación (p.ej., correlación con otra variable ya incluida en el modelo)?
- ¿Es esta variable conceptualmente importante para el análisis salarial, aunque el modelo no encuentre un efecto significativo? ¿Su eliminación cambiaría la interpretación de otros coeficientes?
- ¿Cuál es el costo para el negocio de eliminar esta variable del modelo?

Con base en este análisis, decida qué variables conservar o eliminar, y justifique explícitamente cada decisión. Vuelva a entrenar el modelo con las variables seleccionadas y compare las métricas con el modelo completo.

**Ejemplo del razonamiento esperado:** *"Gender presenta p = 0,18 en nuestro modelo. Sin embargo, optamos por conservarla porque: (1) la equidad salarial por género es una de las preguntas centrales de TalentCo, y (2) su insignificancia puede deberse a colinealidad con Job_Level más que a la ausencia de efecto. Eliminarla ocultaría una dimensión clave del análisis."*

### Paso 2: Análisis de Multicolinealidad (VIF)

Calcule el **Factor de Inflación de la Varianza (VIF)** para todas las variables independientes y responda:

- ¿Qué pares de variables probablemente generan VIF alto, y por qué existe esa correlación en este contexto?
- ¿La presencia de multicolinealidad invalida las predicciones del modelo, o afecta principalmente la confiabilidad de los coeficientes individuales?
- Para cada variable con VIF alto: ¿qué se perdería al eliminarla? ¿Es aceptable esa pérdida dado el objetivo del análisis?

Decida si actuar sobre la multicolinealidad y justifique su decisión. Mantener variables colineales es una respuesta válida si el razonamiento lo sustenta.

**Ejemplo del razonamiento esperado:** *"Age y Years_Experience presentan VIF = 8,4 y 9,1 respectivamente, lo que refleja que los empleados de mayor edad tienden a tener más experiencia. Ambas variables capturan dimensiones distintas —edad biológica vs. tiempo profesional acumulado—, por lo que las conservamos y reconocemos los errores estándar inflados como una limitación del modelo."*

### Paso 3: Verificación de Supuestos de Regresión

Verifique los siguientes supuestos de la regresión lineal sobre su modelo final:

- **Normalidad de los residuos:** Histograma y gráfico Q-Q de los residuos.
- **Homocedasticidad:** Aplique la prueba de Breusch-Pagan (`het_breuschpagan`).

Para cada supuesto, indique si se cumple, muestre la evidencia y discuta brevemente qué implica para la validez de sus conclusiones.



In [37]:
#fixme



## Parte 5: Informe de Negocio

Redacte un informe ejecutivo breve (**300–400 palabras**) dirigido al **Director de Recursos Humanos de TalentCo**, respondiendo las siguientes preguntas:

1. ¿Cuáles son los principales determinantes del salario de los empleados?
2. ¿Cómo impacta cuantitativamente el nivel de experiencia en la compensación anual?
3. ¿Existen diferencias salariales significativas entre departamentos o regiones? ¿Qué sugieren?
4. Según su modelo final: si TalentCo asciende a un empleado de nivel Mid a Senior, ¿cuál es el aumento salarial esperado (en USD)?
5. ¿Qué recomendaciones haría a TalentCo para mejorar la equidad y transparencia en su política de compensaciones?



In [38]:
#fixme


## Entregables

Cada estudiante o grupo debe entregar los siguientes elementos:

| # | Entregable | Descripción |
|---|---|---|
| 1 | **Jupyter Notebook** (`.ipynb`) | Código completo con salidas y comentarios en línea. Debe ser limpio, bien comentado y reproducible. Todos los gráficos deben incluir títulos, etiquetas de ejes y leyendas cuando corresponda. |
| 2 | **Informe de Negocio** (`.md` o `.docx`) | 300 a 400 palabras, dirigido a una audiencia no técnica. |
| 3 | **Presentación Oral** | 10 a 12 minutos (ver rúbrica). |

> ⚠️ **Importante:** Tanto la entrega escrita como la presentación oral son **obligatorias**. La ausencia de cualquiera de los dos componentes resulta en la nota mínima del proyecto.



## Consejos Técnicos

- Utilice las bibliotecas: `pandas`, `numpy`, `matplotlib`, `seaborn`, `scikit-learn` y `statsmodels`.
- Para VIF: `from statsmodels.stats.outliers_influence import variance_inflation_factor`
- Para Breusch-Pagan: `from statsmodels.stats.diagnostic import het_breuschpagan`
- Para gráfico Q-Q: `import scipy.stats as stats; stats.probplot(residuals, plot=plt)`
- Priorice la **interpretación de negocio** por sobre los resultados estadísticos aislados.
- En el informe ejecutivo, escriba para el Director de RRHH: sin fórmulas, sin código, sin p-valores. Solo conclusiones claras y accionables.



## Rúbrica de Presentación Oral

Cada grupo será evaluado en una escala de 10 a 100 puntos. Las presentaciones deben realizarse en español o inglés y tener una duración de entre 10 y 12 minutos.

### Criterios de Evaluación

| Criterio | Descripción | Puntaje | Ponderación |
|---|---|---|---|
| **1. Comprensión del problema y contexto** | Demuestra comprensión clara del problema de negocio de TalentCo y del valor del análisis basado en datos. Vincula los hallazgos estadísticos con decisiones reales de RRHH. | 10–20 pts | 20% |
| **2. Exploración de datos e insights** | Presenta los hallazgos del EDA de forma clara. Destaca patrones relevantes (p.ej., salario por departamento, tendencias de experiencia). Explica cómo el análisis orientó las decisiones de modelado. | 10–20 pts | 20% |
| **3. Explicación del modelo y resultados** | Explica los modelos de regresión, las decisiones de codificación y las métricas de evaluación (MAE, RMSE, R²) con precisión. Compara OLS vs. sklearn y justifica el modelo final. | 10–20 pts | 20% |
| **4. Interpretación de negocio y recomendaciones** | Traduce los coeficientes del modelo y los resultados de las pruebas en recomendaciones accionables para RRHH. Explica qué debe hacer TalentCo y por qué, sin jerga técnica. | 10–20 pts | 20% |
| **5. Comunicación y presentación** | Exposición clara y profesional en español o inglés. Flujo lógico, tono seguro y trabajo en equipo efectivo. Respeta el tiempo establecido. | 10–15 pts | 15% |
| **6. Apoyos visuales y diseño de diapositivas** | Las diapositivas son claras, profesionales y apoyan la narrativa. Los gráficos y tablas son legibles y están bien etiquetados. Evita diapositivas con exceso de texto. | 10–15 pts | 15% |

### Escala de Calificación

| Puntaje | Nivel | Descripción |
|---|---|---|
| 90–100 | ⭐ Sobresaliente | Comprensión excepcional, insights de negocio sólidos y presentación atractiva. Vínculo claro entre el análisis de datos y la estrategia de RRHH. |
| 80–89 | ✅ Muy Bueno | Explicación y visuales sólidos, con brechas menores en profundidad o fluidez. Buen dominio técnico y de negocio. |
| 70–79 | ⚙️ Bueno | Cubre los puntos principales de manera adecuada, pero con falta de profundidad o claridad en algunas áreas. Estructura razonable. |
| 60–69 | ⚠️ Aceptable | Cobertura básica con insight de negocio limitado o conexión débil con el problema de RRHH. Presentación algo confusa o apresurada. |
| 10–59 | ❌ Insuficiente | Estructura deficiente, contenido incompleto o incapacidad para comunicar los hallazgos. Puede exceder o quedar muy por debajo del tiempo establecido. |

### Notas para Evaluadores

- Descontar hasta **−10 puntos** si la presentación no se encuentra dentro del rango de tiempo (menos de 7 min o más de 13 min).
- Descontar hasta **−5 puntos** si no todos los integrantes participan activamente.
- Otorgar hasta **+5 puntos de bonificación** por insight excepcional, creatividad o conexión con casos reales.

**Puntaje máximo:** 100 puntos | **Nota mínima de aprobación:** 60 puntos
